psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java

```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
 		return inc * groundRule.getWeight();
 	}
 
+	/** Compute the rounding probability function for a given truth value */
+	public static double roundingProbability(Double truthValue) {
+		return 0.25 + 0.5 * truthValue;
+	}
+
+	/**
+	 * Computes the expected weighted compatibility (1 - incompatibility)
+	 * of a WeightedGroundLogicalRule.
+	 * from independently rounding each {@link RandomVariableAtom} to 1.0 or 0.0
+	 * with probability equal to its current truth value.
+	 *
+	 * WARNING: the result of this function is incorrect if the RandomVariableAtoms
+	 * are subject to any {@link UnweightedGroundRule UnweightedGroundRule}.
+	 *
+	 * WARNING: This method does not account for WeightedGroundLogicalRules that
+	 * were not grounded because they are trivially satisfied.
+	 */
+	public static double getExpectedWeightedLogicalCompatibility(WeightedGroundLogicalRule groundRule) {
+		/* Collects GroundAtom values. Atoms in GroundRule are
+		 * already negated */
+		List<Double> posRvValues = new ArrayList<Double>();
+		List<Double> posObsValues = new ArrayList<Double>();
+		List<Double> negRvValues = new ArrayList<Double>();
+		List<Double> negObsValues = new ArrayList<Double>();
+		for (GroundAtom atom : groundRule.getPositiveAtoms())
+			if (atom instanceof RandomVariableAtom)
+				negRvValues.add(atom.getValue());
+			else
+				negObsValues.add(atom.getValue());
+		for (GroundAtom atom : groundRule.getNegativeAtoms())
+			if (atom instanceof RandomVariableAtom)
+				posRvValues.add(atom.getValue());
+			else
+				posObsValues.add(atom.getValue());
+
+		// Sum of observed atoms
+		double k = 0;
+		for (double posObsValue : posObsValues)
```

Comment:
Do this work in the loops above, then we don't have to store the constants in lists.